### PR TITLE
Integrate MFEM's CUDA support into serac + AMGX, merge in changes from MFEM spack package

### PR DESCRIPTION
**Pull request by joshessman-llnl**
This PR now builds `mfem+cuda` when Serac's `+cuda` variant is enabled, and adds a new `accelerator` "component" that wraps MFEM's `Device` object.

Actual performance improvements are probably negligible at best as the relevant integrators used in the driver are not (to my knowledge) GPU-enabled.  It would probably be worthwhile to test this with a large mesh if more precise information is desired.

### EDIT

This PR also adds an AMGX preconditioner option (and an associated test).  This required a version bump to MFEM 4.2.

### CODE REVIEW

File: cmake/thirdparty/FindMFEM.cmake
Comment (white238) [inline cmake/thirdparty/FindMFEM.cmake]: Could you not use the BLT CUDA registered libraries? `cuda` or `cuda_runtime`

File: host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@upstream_gfortran-cuda.cmake
Comment (white238) [inline host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@upstream_gfortran-cuda.cmake]: was this reverted on purpose?

Comment (joshessman-llnl) [reply host-configs/lassen-blueos_3_ppc64le_ib_p9-clang@upstream_gfortran-cuda.cmake]: It was actually an oversight when I was juggling images around - the change to our version of axom's package.py is actually in https://github.com/LLNL/serac/pull/88/files#diff-f2386102a0ae90f37dd97db98adcfd7e856a8bf99811ffe7649e55245e2368e9R58 (which hasn't been merged).

That PR needs to be merged in before this one, so there will be a rebuild with the correct tag name.

File: cmake/thirdparty/FindMFEM.cmake
Comment (joshessman-llnl) [reply cmake/thirdparty/FindMFEM.cmake]: I'm not sure that BLT-registered libraries can be used when registering a BLT library - I could be wrong though.

Either way, the runtime libraries alone are not sufficient because MFEM also requires `cusparse` but doesn't reference it in the `MFEM_EXT_LIBS` variable that we use to get the paths to the rest of its libraries.

Comment (white238) [reply cmake/thirdparty/FindMFEM.cmake]: You can, just list it on the LIBRARIES line.  You should also be able to just add `cusparse`.

Comment (white238) [reply cmake/thirdparty/FindMFEM.cmake]: Just to correct my previous statement... blt registered libraries should go on the DEPENDS_ON line.  /hangHeadInShame

**Comment by joshessman-llnl** [comment]
This won't work properly until Axom can be built against a version of MFEM built with CUDA support.

File: scripts/uberenv/packages/mfem/package.py
Comment (jamiebramwell) [inline scripts/uberenv/packages/mfem/package.py]: We should pull in the official tarball version of 4.2.0 from their spack package. These package updates should also probably get incorporated: https://github.com/spack/spack/pull/19748 .

File: cmake/thirdparty/FindMFEM.cmake
Comment (joshessman-llnl) [reply cmake/thirdparty/FindMFEM.cmake]: This will be handled in a cleaner way once https://github.com/LLNL/blt/pull/421 is merged

File: scripts/uberenv/packages/mfem/package.py
Comment (joshessman-llnl) [reply scripts/uberenv/packages/mfem/package.py]: The Spack changes have been merged in, and once this PR is merged I can open a PR to add the AMGX integration to the official Spack package (which isn't part of that PR)

**Comment by joshessman-llnl** [comment]
LGTM

File: src/physics/utilities/solver_config.hpp
Comment (white238) [inline src/physics/utilities/solver_config.hpp:88]: Why is this just an empty struct?

**Review by white238 — APPROVED** [review]

File: src/physics/utilities/solver_config.hpp
Comment (joshessman-llnl) [reply src/physics/utilities/solver_config.hpp:88]: Right now it's only being used as the "tag" in the preconditioner variant, as the default settings appeared to be sufficient for the test I added, but in the future I think we'll want to configure AMGX as we do Hypre's AMG preconditioner implementation.

**Comment by joshessman-llnl** [comment]
@jamiebramwell Now that the "official" 4.2 updates to our version of the MFEM spack package have been merged in, was there anything else you were looking to see as part of this PR?

**Comment by jamiebramwell** [comment]
I'll give this another look, but I think it should be good to go as long as the TPLs are built on LC and in docker.

File: host-configs/docker/docker-linux-ubuntu16.04-x86_64-gcc@8.1.0.cmake
Comment (jamiebramwell) [inline host-configs/docker/docker-linux-ubuntu16.04-x86_64-gcc@8.1.0.cmake:62]: 👍

**Review by jamiebramwell — APPROVED** [review]

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Comment by joshessman-llnl** [comment]
LGTM

**Pull request merged**
